README.md
# low-disc

A Clojure library of [low discrepancy
sequence](https://en.wikipedia.org/wiki/Low-discrepancy_sequence) algorithms.

The following sequences are implemented:

* Van der Corput sequence
* Halton sequence
* Hammersley sequence
* Uniform random sequence (for comparison)

## Installation

Latest version: 0.1.0

Leiningen dependency information:

```Clojure
[low-disc "0.1.0"]
```

## Illustrations

### Uniform

![](/img/uniform.png)

### Halton Sequence

| Dimensions  | Unscrambled | Scrambled |
|----------------------|:-----------:|:---------:|
| 1 and 2   | ![](/img/halton-unscr.png) | ![](/img/halton.png) |
| 15 and 16 | ![](/img/halton-unscr-high.png) | ![](/img/halton-high.png) |

### Hammersley Sequence


| Dimensions  | Unscrambled | Scrambled |
|----------------------|:-----------:|:---------:|
| 1 and 2   | ![](/img/hammersley-unscr.png) | ![](/img/hammersley.png) |
| 15 and 16 | ![](/img/hammersley-unscr-high.png) | ![](/img/hammersley-high.png) |


## Usage

All sequence functions return lazy sequences of points. Points are `n`-dimensional vectors whose
coordinates are each between 0 and 1. The only exception is the one dimensional van der Corput
sequence whose elements are scalars.

With the exception of sequences that are finite by definition, all sequence functions return
lazy inifinite sequences by default. If the optional parameter `l` is given, only the first `l`
elements are returned.

Many sequences come in two variants: scrambled and unscrambled ones. The scrambled one is usually
the default, whereas the unscrambled one is denoted by the suffix `unscr` (e.g. `halton-2d` vs.
`halton-unscr-2d`).

### Van der Corput Sequence

The van der Corput sequence is parametrized by a basis `b` which should be a prime number. It can be
scrambled using a permutation `p`, specified in the form of a vector: It must contain the first `b`
integers starting with `0` in an order which defines the permutation (e.g. for basis `3` `[0 2 1]`)

```Clojure
(vdc p b)
(vdc p b l)

(vdc-unscr b)
(vdc-unscr b l)
```

```Clojure
(vdc-unscr 3 5)  ;;=> (1/3 2/3 1/9 4/9 7/9)
```

### Halton Sequence

The Halton sequence is a generalization of the van der Corput sequence to the n-dimensional case.
Consequently, it is parametrized by `n` prime bases which should be distinct. It can be scrambled
by `n` permutations. 

Permutations can be specified manually using the `halton` function. Alternatively, the `halton-nd`
functions provide defaults for the first 16 dimensions taken from the literature.

```Clojure
(halton ps bs)
(halton ps bs l)

(halton-unscr bs)
(halton-unscr bs l)

(halton-nd n)
(halton-nd n l)

(halton-unscr-nd n)
(halton-unscr-nd n l)

(halton-1d)
(halton-1d l)
...
(halton-16d)
(halton-16d l)

(halton-unscr-1d)
(halton-unscr-1d l)
...
(halton-unscr-16d)
(halton-unscr-16d l)
```

```Clojure
(halton-2d 3)  ;;=> ([1/2 2/3] [1/4 1/3] [3/4 2/9])
```

### Hammersley sequence

The Hammersley sequence has a defined size `l`. There are both scrambled and unscrambled versions.
The former use the same defaults as their Halton equivalents.

```Clojure
(hammersley-nd n l)

(hammersley-unscr-nd n l)

(hammersley-1d l)
...
(hammersley-16d l)

(hammersley-unscr-1d l)
...
(hammersley-unscr-16d l)
```

```Clojure
(hammersley-2d 3)  ;;=> ([1/2 0] [1/4 1/3] [3/4 2/3])
```

### Random sequence

For comparison, random sequences are provided. They consist of points chosen uniformly, at random,
and independently of each other inside the unit plane (cube, ...).

```Clojure
(uniform-nd n)
(uniform-nd n l)

(uniform-1d)
(uniform-1d l)
...
(uniform-16d)
(uniform-16d l)
```

```Clojure
(uniform-2d 3) ;;=> ([0.5270699200579589 0.47777654304867256]
               ;;    [0.8942239443816675 0.032992693354966796]
               ;;    [0.7448379865121517 0.4345372587047025])
```

## License

Copyright © 2020 Jannik Luhn

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 0.1.0 - 2020-04-11
### Added
- Van der Corput sequence.
- Halton sequence.
- Hammersley sequence.
- Random sequence.


test/low_disc/test.clj
(ns low-disc.test
  (:require [clojure.test :refer :all]
            [low-disc.utils :refer [digits-rev]]
            [low-disc.core :refer :all]))


(deftest utils-tests
  (is (= (digits-rev 10 123405) [5 0 4 3 2 1]))
  (is (= (digits-rev 10 0) [])))

(deftest vdc-tests
  (is (= (take 4 (vdc-unscr 2)) [1/2 1/4 3/4 1/8]))
  (is (every? #(< 0 % 1) (take 1000 (vdc-unscr 29))))
  (is (every? #(< 0 % 1) (take 1000 (vdc [0 3 1 4 2] 5))))
  (is (= (take 4 (vdc-unscr 2)) (vdc-unscr 2 4))))

(deftest halton-tests
  (is (= (take 4 (halton-unscr [2 3])) [[1/2 1/3] [1/4 2/3] [3/4 1/9] [1/8 4/9]]))
  (let [h (take 1000 (halton [[0 2 1] [0 4 2 3 1]] [3 5]))]
    (is (every? (fn [[x y]] (and (< 0 x 1) (< 0 y 1))) h))
    (is (every? #(= (count %) 2) h)))
  (is (= (count (first (halton-nd 1))) 1))
  (is (= (count (first (halton-nd 2))) 2))
  (is (= (count (first (halton-nd 15))) 15))
  (is (= (take 4 (halton-4d)) (halton-4d 4))))

(deftest hammersley-tests
  (is (= (count (hammersley-5d 100)) 100))
  (is (every? #(= % 5) (map count (hammersley-5d 100)))))


src/low_disc/plot.clj
(ns low-disc.plot
  (:require [quil.core :as q]
            [low-disc.core :as ld]))

(def width 500)
(def height 500)
(def n 500)
(def base-hue 0.25)

(defn draw-marker [x y]
  (q/ellipse (* x width) (* y height) 10 10))

(defn draw-seq [s]
  (let [[s1 s2] (split-at (quot (count s) 10) s)]
    (q/fill base-hue 1 1)
    (doseq [[x y] s2] (draw-marker x y))
    (q/fill (rem (+ base-hue 0.5) 1) 1 1)
    (doseq [[x y] s1] (draw-marker x y))))

(defn draw []
  (q/no-loop)
  (q/ellipse-mode :center)
  (q/color-mode :hsb 1)
  (q/background 1)
  (q/no-stroke)
  (draw-seq (take n (ld/halton-unscr-2d)))
  #_(draw-seq (map (partial take-last 2) (ld/hammersley-16d n)))
  #_(draw-seq (take n (ld/uniform-2d)))
  #_(q/save-frame "halton-unscr.png"))

; run sketch
(q/defsketch trigonometry
  :size [width height]
  :draw draw
  :middlewares [:no-bind-output])


src/low_disc/halton.clj
(ns low-disc.halton
  (:require [low-disc.vdc :as vdc]))

(def ^:private primes [2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53])

(def ^:private permutations
  "from https://doi.org/10.1016/j.cam.2005.05.022"
  {2 [0 1]
   3 [0 2 1]
   5 [0 2 4 1 3]
   7 [0 3 5 1 6 2 4]
   11 [0 5 8 2 10 3 6 1 9 4 7]
   13 [0 6 10 2 8 4 12 1 9 5 11 3 7]
   17 [0 8 13 3 11 5 16 1 10 7 14 4 12 2 15 6 9]
   19 [0 9 14 3 17 6 11 1 15 7 12 4 18 8 2 16 10 5 13]
   23 [0 11 17 4 20 7 13 2 22 9 15 5 18 1 14 10 21 6 16 3 19 8 12]
   29 [0 14 22 5 18 9 27 2 20 11 25 7 16 3 24 13 19 6 28 10 1 23 15 12 26 4 17 8 21]
   31 [0 16 8 26 4 22 13 29 2 19 11 24 6 20 14 28 1 17 9 30 10 23 5 21 15 3 27 12 25 7 18]
   37 [0 18 28 6 23 11 34 3 25 14 31 8 20 36 1 16 27 10 22 13 32 4 29 17 7 35 19 2 26 12 30 9 24 15
       33 5 21]
   41 [0 20 31 7 26 12 38 3 23 34 14 17 29 5 40 10 24 1 35 18 28 9 33 15 21 4 37 13 30 8 39 19 25 2
       32 11 22 36 6 27 16]
   43 [0 21 32 7 38 13 25 3 35 17 28 10 41 5 23 30 15 37 1 19 33 11 26 42 8 18 29 4 39 14 22 34 6
       24 12 40 2 31 20 16 36 9 27]
   47 [0 23 35 8 41 14 27 3 44 18 31 11 37 5 25 39 16 21 33 1 46 12 29 19 42 7 28  10 36 22 4 43 17
       32 13 38 2 26 45 15 30 6 34 20 40 9 24]
   53 [0 26 40 9 33 16 49 4 36 21 45 12 29 6 51 23 38 14 43 1 30 19 47 10 34 24 42 3 27 52 15 18 39
       7 46 22 32 5 48 13 35 25 8 44 31 17 50 2 37 20 28 11 41]})

(defn halton
  "Returns the Halton sequence for bases bs scrambled with permutations ps."
  ([ps bs]
   (apply map vector (map vdc/vdc ps bs)))
  ([ps bs l]
   (take l (halton))))

(defn halton-nd
  "Returns an n dimensional scrambled Halton sequence.

  Without an explicitly given length l, an infinite sequence is created

  The first n prime numbers act as bases.

  Scrambling is performed using the permutations found using Braaten and Weller's algorithm by
  Vandewoestyne and Cools (https://doi.org/10.1016/j.cam.2005.05.022)."
  ([n]
   {:pre [(<= n (max (count permutations)
                    (count primes)))]}
   (let [bs (take n primes)
         ps (map permutations bs)]
     (halton ps bs)))
  ([n l] (take l (halton-nd n))))

(defn halton-unscr
  "Returns the unscrambled Halton sequence for bases bs.

  Without an explicitly given length l, an infinite sequence is created."
  ([bs]
   (halton (map range bs) bs))
  ([bs l]
   (take l (halton-unscr bs))))

(defn halton-unscr-nd
  "Returns an n dimensional unscrambled Halton sequence.

  Without an explicitly given length l, an infinite sequence is created

  The first n prime numbers act as bases."
  ([n]
   {:pre [(<= n (count primes))]}
   (halton-unscr (take n primes) n))
  ([n l] (take l (halton-unscr-nd n))))

; define halton-1d and halton-unscr-1d up until halton-16d and halton-unscr-16d
(dotimes [n 16]
  (intern *ns*
          (symbol (str "halton-" n "d"))
          (partial halton-nd n))
  (intern *ns*
          (symbol (str "halton-unscr-" n "d"))
          (partial halton-nd n)))


src/low_disc/core.clj
(ns low-disc.core
  (:require [low-disc.utils :refer [digits-rev]]))

;
; Van der Corput
;
(defn vdc-el [p b n]
  "Returns the nth element of the base b van der Corput sequence scrambled with permutation p."
  (->> (map #(/ %1 %2)
            (map #(nth p %) (digits-rev b n))
            (iterate #(* b %) b))
       (reduce +)))

(defn vdc
  "Returns the base b van der Corput sequence scrambled with permutation p.

  Without an explicitly given length l, an infinite sequence is created."
  ([p b]
   (->> (range)
        (map inc)
        (map #(vdc-el p b %))))
  ([p b l]
   (take l (vdc p b))))

(defn vdc-unscr
  "Returns the unscrambled base b van der Corput sequence.

  Without an explicitly given length l, an infinite sequence is created."
  ([b] (vdc (range b) b))
  ([b l] (take l (vdc-unscr b))))

;
; Halton
;
(def ^:private primes [2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53])

(def ^:private permutations
  "from https://doi.org/10.1016/j.cam.2005.05.022"
  {2 [0 1]
   3 [0 2 1]
   5 [0 2 4 1 3]
   7 [0 3 5 1 6 2 4]
   11 [0 5 8 2 10 3 6 1 9 4 7]
   13 [0 6 10 2 8 4 12 1 9 5 11 3 7]
   17 [0 8 13 3 11 5 16 1 10 7 14 4 12 2 15 6 9]
   19 [0 9 14 3 17 6 11 1 15 7 12 4 18 8 2 16 10 5 13]
   23 [0 11 17 4 20 7 13 2 22 9 15 5 18 1 14 10 21 6 16 3 19 8 12]
   29 [0 14 22 5 18 9 27 2 20 11 25 7 16 3 24 13 19 6 28 10 1 23 15 12 26 4 17 8 21]
   31 [0 16 8 26 4 22 13 29 2 19 11 24 6 20 14 28 1 17 9 30 10 23 5 21 15 3 27 12 25 7 18]
   37 [0 18 28 6 23 11 34 3 25 14 31 8 20 36 1 16 27 10 22 13 32 4 29 17 7 35 19 2 26 12 30 9 24 15
       33 5 21]
   41 [0 20 31 7 26 12 38 3 23 34 14 17 29 5 40 10 24 1 35 18 28 9 33 15 21 4 37 13 30 8 39 19 25 2
       32 11 22 36 6 27 16]
   43 [0 21 32 7 38 13 25 3 35 17 28 10 41 5 23 30 15 37 1 19 33 11 26 42 8 18 29 4 39 14 22 34 6
       24 12 40 2 31 20 16 36 9 27]
   47 [0 23 35 8 41 14 27 3 44 18 31 11 37 5 25 39 16 21 33 1 46 12 29 19 42 7 28  10 36 22 4 43 17
       32 13 38 2 26 45 15 30 6 34 20 40 9 24]
   53 [0 26 40 9 33 16 49 4 36 21 45 12 29 6 51 23 38 14 43 1 30 19 47 10 34 24 42 3 27 52 15 18 39
       7 46 22 32 5 48 13 35 25 8 44 31 17 50 2 37 20 28 11 41]})

(defn halton
  "Returns the Halton sequence for bases bs scrambled with permutations ps."
  ([ps bs]
   (let [vdcs (map vdc ps bs)]
     (if (seq vdcs)
       (apply map vector vdcs)
       (repeat []))))
  ([ps bs l]
   (take l (halton ps bs))))

(defn halton-nd
  "Returns an n-dimensional scrambled Halton sequence.

  Without an explicitly given length l, an infinite sequence is created

  The first n prime numbers act as bases.

  Scrambling is performed using the permutations found using Braaten and Weller's algorithm by
  Vandewoestyne and Cools (https://doi.org/10.1016/j.cam.2005.05.022)."
  ([n]
   {:pre [(<= n (max (count permutations)
                    (count primes)))]}
   (let [bs (take n primes)
         ps (map permutations bs)]
     (halton ps bs)))
  ([n l] (take l (halton-nd n))))

(defn halton-unscr
  "Returns the unscrambled Halton sequence for bases bs.

  Without an explicitly given length l, an infinite sequence is created."
  ([bs]
   (halton (map range bs) bs))
  ([bs l]
   (take l (halton-unscr bs))))

(defn halton-unscr-nd
  "Returns an n-dimensional unscrambled Halton sequence.

  Without an explicitly given length l, an infinite sequence is created

  The first n prime numbers act as bases."
  ([n]
   {:pre [(<= n (count primes))]}
   (halton-unscr (take n primes)))
  ([n l] (take l (halton-unscr-nd n))))

; halton-1d ... halton-16d, halton-unscr-1d ... halton-unscr-16d
(doseq [n (range 1 17)]
  (intern *ns*
          (symbol (str "halton-" n "d"))
          (partial halton-nd n))
  (intern *ns*
          (symbol (str "halton-unscr-" n "d"))
          (partial halton-unscr-nd n)))

;
; Hammersley
;
(defn hammersley-nd [n l]
  "Return the scrambled n-dimensional Hammersley set of size l.

  The implementation is based on the scrambled Halton sequence."
  (->> (halton-nd (dec n))
       (map-indexed #(conj %2 (/ %1 l)))
       (take l)))

(defn hammersley-unscr-nd [n l]
  "Return the unscrambled n-dimensional Hammersley set of size l.

  The implementation is based on the unscrambled Halton sequence."
  (->> (halton-unscr-nd (dec n))
       (map-indexed #(conj %2 (/ %1 l)))
       (take l)))

; hammersley-1d ... hammersley-16d, hammersley-unscr-1d ... hammersley-16d
(doseq [n (range 1 17)]
  (intern *ns*
          (symbol (str "hammersley-" n "d"))
          (partial hammersley-nd n))
  (intern *ns*
          (symbol (str "hammersley-unscr-" n "d"))
          (partial hammersley-unscr-nd n)))

;
; Uniform
;
(defn uniform-nd-point [n]
  "Return a single n-dimensional point chosen uniformly at random."
  (into [] (repeatedly n rand)))

(defn uniform-nd
  "Return a lazy sequence of n-dimensional points chosen independently and uniformly at random.

  Without an explicitly given length l, an infinite sequence is created."
  ([n] (repeatedly (partial uniform-nd-point n)))
  ([n l] (take l (uniform-nd n))))

; uniform-1d ... uniform-16d
(doseq [n (range 1 17)]
  (intern *ns*
          (symbol (str "uniform-" n "d"))
          (partial uniform-nd n)))
